utils/db_utils.py

```diff
@@ -158,4 +158,104 @@ def add_user_stats(user_data):
+    
+    vals = []
+    user_id_seen_dict = {}
+    number_user_seen = 0
+    # Now we can read every section crossing start point
+    for section_id in start_by_section:
+        matching_start = start_by_section[section_id]
+        matching_start_data = matching_start.get('data')
+        if matching_start_data is None:
+            # Something is wrong with the fetched data, shouldn't happen
+            continue
+        if 'idx' not in matching_start_data:
+            # Sometimes, idx is missing in data, not sure why
+            continue
+        matching_end = end_by_section.get(section_id)
+        if matching_end is None:
+            # This section_id didn't cross the end point
+            continue
+        matching_end_data = matching_end.get('data', {})
+        # idx allows us to check that the start section is crossed first, we do not care about trips going the other way
+        if 'idx' in matching_end_data and matching_start_data.get('idx') < matching_end_data.get('idx'):
+            user_id = str(start_by_section[section_id].get('user_id'))
+            if user_id_seen_dict.get(user_id) is None:
+                number_user_seen += 1
+                user_id_seen_dict[user_id] = True
+            vals.append({
+                'start': start_by_section[section_id], 
+                'end': end_by_section[section_id], 
+                'duration': matching_end_data.get('ts') - matching_start_data.get('ts'),
+                'section': section_id, 
+                'mode': matching_start_data.get('mode'), # Note: this is the mode given by the phone, not the computed one, we'll read it later from inferred_sections
+                'start_fmt_time': matching_start_data.get('fmt_time'),
+                'end_fmt_time': matching_end_data.get('fmt_time')
+            })
+    if perm_utils.permissions.get("segment_trip_time_min_users", 0) <= number_user_seen:
+        return pd.DataFrame.from_dict(vals)
+    return pd.DataFrame.from_dict([])
+
+# The following query can be called multiple times, let's open db only once
+analysis_timeseries_db = edb.get_analysis_timeseries_db()
+
+# When sections isn't set, this fetches all inferred_section
+# Otherwise, it filters on given section ids using '$in'
+# Note: for performance reasons, it is not recommended to use '$in' a list bigger than ~100 values
+# In our use case, this could happen on popular trips, but the delay is deemed acceptable
+def query_inferred_sections_modes(sections=[]):
+    query = {'metadata.key': 'analysis/inferred_section'}
+    if len(sections) > 0:
+        query['data.cleaned_section'] = {'$in': sections} 
+    res = analysis_timeseries_db.find(query, {'data.cleaned_section': 1, 'data.sensed_mode': 1})
+    mode_by_section_id = {}
+    for elt in res:
+        elt_data = elt.get('data')
+        if elt_data:
+            mode_by_section_id[str(elt_data.get('cleaned_section'))] = elt_data.get('sensed_mode') or 0
```

Comment:
I see your note here about performance issues while reading the inferred sections.
And I also see that you are making a direct mongoDB query from here, which we want to avoid  to make it easier to refactor the core as needed later.

I would suggest iterating through the sections and using `cleaned2inferred_section` in `emission/storage/decorations/section_queries` on each one.

As an aside, I should really think about how to fix the datamodel so that we don't need this two step retrieval and the unnecessary performance hit.

Reply:
I have filed https://github.com/e-mission/e-mission-docs/issues/970 to create a new interface for this. Once we have the interface, we can experiment or switch implementations under the hood. My personal preference is for the timeseries model (so we find potentially matching inferred sections by querying by time range or start/end geo) but I don't know if that will be feasible for this use case - e.g. whether the query will be tight enough.